examples/conway/main.rs

```diff
@@ -0,0 +1,394 @@
+            c.cool_off(0.4);
+        }
+    }
+
+    fn count_neibs(&self, x: usize, y: usize) -> usize {
+        let mut count = 0;
+        let wi = self.width as isize;
+        let hi = self.height as isize;
+        for j in -1isize..=1 {
+            for i in -1isize..=1 {
+                if i == 0 && j == 0 {
+                    continue;
+                }
+                // wrap around
+                let xx = (x as isize + i + wi) % wi;
+                let yy = (y as isize + j + hi) % hi;
+                assert!(xx >= 0 && yy >= 0 && xx < wi && yy < hi);
+                let i = self.grid_idx(xx, yy).expect("wrap-around bug?");
+                if self.cells[i].alive {
+                    count += 1;
+                }
+            }
+        }
+        count
+    }
+
+    fn update(&mut self) {
+        for y in 0..self.height {
+            for x in 0..self.width {
+                let neibs = self.count_neibs(x, y);
+                let idx = x + y * self.width;
+                let next = self.cells[idx].update_neibs(neibs);
+                // Write into scratch_cells, since we're still reading from `self.cells`
+                self.scratch_cells[idx] = next;
+            }
+        }
+        std::mem::swap(&mut self.scratch_cells, &mut self.cells);
+    }
+
+    fn toggle(&mut self, x: isize, y: isize) -> bool {
+        if let Some(i) = self.grid_idx(x, y) {
+            let was_alive = self.cells[i].alive;
+            self.cells[i].set_alive(!was_alive);
+            !was_alive
+        } else {
+            false
+        }
+    }
+
+    fn draw(&self, screen: &mut [u8]) {
+        debug_assert_eq!(screen.len(), 4 * self.cells.len());
+        for (c, pix) in self.cells.iter().zip(screen.chunks_exact_mut(4)) {
+            let color = if c.alive {
+                [0, 0xff, 0xff, 0xff]
+            } else {
+                [0, 0, c.heat, 0xff]
+            };
+            pix.copy_from_slice(&color);
+        }
+    }
+
+    fn set_line(&mut self, x0: isize, y0: isize, x1: isize, y1: isize, alive: bool) {
+        // probably should do sutherland-hodgeman if this were more serious.
+        // instead just clamp the start pos, and draw until moving towards the
+        // end pos takes us out of bounds.
+        let x0 = x0.max(0).min(self.width as isize);
+        let y0 = y0.max(0).min(self.height as isize);
+        bresenham(x0, y0, x1, y1, |x, y| {
+            if let Some(i) = self.grid_idx(x, y) {
+                self.cells[i].set_alive(alive);
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    fn grid_idx<I: std::convert::TryInto<usize>>(&self, x: I, y: I) -> Option<usize> {
+        if let (Ok(x), Ok(y)) = (x.try_into(), y.try_into()) {
+            Some(x + y * self.width)
+        } else {
+            None
+        }
+    }
+}
+
+fn bresenham(x0: isize, y0: isize, x1: isize, y1: isize, mut cb: impl FnMut(isize, isize) -> bool) {
```

Comment:
I honestly considered doing this, too. But took a shortcut when I found the [`line_drawing`](https://docs.rs/line_drawing/0.8.0/line_drawing/struct.Bresenham.html) crate, which is already included as a dev-dependency. Maybe worth just reusing that?

Reply:
Sure. I didn't notice the dep existed, and wasn't about to add a new one for a short function.